quickwit/quickwit-indexing/src/source/queue_sources/visibility.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use quickwit_actors::{
+    Actor, ActorContext, ActorExitStatus, ActorHandle, ActorState, Handler, Mailbox,
+};
+use serde_json::{json, Value as JsonValue};
+
+use super::Queue;
+use crate::source::SourceContext;
+
+#[derive(Debug, Clone)]
+pub(super) struct VisibilitySettings {
+    /// The original deadline asked from the queue when polling the messages
+    pub deadline_for_receive: Duration,
+    /// The last deadline extension when the message reading is completed
+    pub deadline_for_last_extension: Duration,
+    /// The extension applied why the VisibilityTask to maintain the message visibility
+    pub deadline_for_default_extension: Duration,
+    /// Rhe timeout for the visibility extension request
+    pub request_timeout: Duration,
+    /// an extra margin that is substracted from the expected deadline when
+    /// asserting whether we are still in time to extend the visibility
+    pub request_margin: Duration,
+}
+
+impl VisibilitySettings {
+    /// The commit timeout gives us a first estimate on how long the processing
+    /// will take for the messages. We could include other factors such as the
+    /// message size.
+    pub(super) fn from_commit_timeout(commit_timeout_secs: usize) -> Self {
+        let commit_timeout = Duration::from_secs(commit_timeout_secs as u64);
+        Self {
+            deadline_for_receive: Duration::from_secs(120) + commit_timeout,
+            deadline_for_last_extension: 2 * commit_timeout,
+            deadline_for_default_extension: Duration::from_secs(60),
+            request_timeout: Duration::from_secs(3),
+            request_margin: Duration::from_secs(1),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct VisibilityTask {
+    queue: Arc<dyn Queue>,
+    ack_id: String,
+    extension_count: u64,
+    current_deadline: Instant,
+    stop_extension_loop: bool,
+    visibility_settings: VisibilitySettings,
+}
+
+// A handle to the visibility actor. When dropped, the actor exits and the
```

Comment:
How does that work? is this tested?

Reply:
My understanding is that dropping the last mailbox of an actor exits it: https://github.com/quickwit-oss/quickwit/blob/207626185a5327d2d6fd84181f3e383fd274fe57/quickwit/quickwit-actors/src/mailbox.rs#L49-L50

But after adding some tests in https://github.com/quickwit-oss/quickwit/pull/5148/commits/95beb16ce36efae2bea5ae36e97b3495796e39d4 I realized that wasn't the case.

Reply:
added an `Arc<()>` to perform ref counting instead of relying on the mailbox